awx/main/models/credential/__init__.py

```diff
@@ -363,6 +372,14 @@ def passwords_needed(self):
             else:
                 needed.append('vault_password')
         return needed
+    
+    @property
+    def dynamic_input_fields(self):
+        dynamic_input_fields = getattr(self, '_dynamic_input_fields', None)
+        if dynamic_input_fields is None:
+            self._dynamic_input_fields = self._get_dynamic_input_field_names()
+            return self._dynamic_input_fields
+        return dynamic_input_fields
```

Comment:
The way that you're handing the internal property looks like this is a vanilla cached property. You might want to have a look at `from django.utils.functional import cached_property`, it's just a way to avoid having to maintain that overhead yourself. But this done in a million places, and its current form is fine with me.

Thanks for getting rid of the overrides of `__init__` and other methods like that, this is looking much better.

Reply:
Neat! This reminds me of the classic cache decorator [covered in a really great pycon talk a few years back](https://speakerdeck.com/pyconslides/transforming-code-into-beautiful-idiomatic-python-by-raymond-hettinger-1?slide=33). I thought about adding something similar for this, but I didn't want to add a new helper function for a single use case.

As you've pointed, django provides something comparable its library of utility functions so we can just use that 👍